### ISSUE TITLE
[DependencyInjection] Request for anonymous function factories

### ISSUE THREAD

**Issue by unkind** [symfony/symfony#28992]
It was already proposed few years ago: #12115, but I want to revisit this issue again.

I'm OK if you still don't accept 
```php
return function (ContainerConfigurator $c) {
    $s = $c->services()
        ->defaults()
        ->autoconfigure(true)
        ->autowire(true)
        ->private();

    $s->set('doctrine_pdo', \Doctrine\DBAL\Connection::class)
        ->factory(function () {
            return DriverManager::getConnection(...);
        });

};
```
, but I want it at least for my projects.

Major issues with this proposal were:
1. `use` may hold reference to object and cannot be inlined to the generated static factory.
2. it might be actually _closed by_ (to be a _closure_) some object (`$this`) and unfortunately, PHP documentation makes no difference between anonymous functions and closures. But this one could be solved by requirement to prefix anonymous functions with "static" and throw exception in case of violation:
   ```
    ->factory(
        static function (...) {
            // ...
        }
    )
   ```
3. this one by @stof, but I still don't get it, probably it was about inlining?
  > Btw, I see a big drawback with this approach: any time your factory references another of its methods to get a dependency, it will create its own instance of the services, which will not be the same than the shared service used by the container. 
4. It required to have dependency on `SuperClosure`. 
 
But since you have `symfony/contracts` repo, I think we can add something like `Symfony\Closures\ClosureDumper` interface and make corresponding changes in the `PhpDumper`. It allows you to avoid dependency on 3rd party libraries like `SuperClosure`, but opens extension points for users like me:
```php
namespace Symfony\Component\BoostedReflection;

interface AnonymousFunctionDumper
{
    /**
     * @throws AnonymousFunctionScopeContainsReferences foo () use (&a, $object)
     * @throws AnonymousFunctionIsClosureForReal Holds reference to $this
     *                    i.e. is not an anonymous function, strictly speaking
     * @return PHP code
     */
    public function dump(\Closure $closure): string;
}
```

btw, `Symfony\Component\DependencyInjection\LazyProxy\Instantiator\InstantiatorInterface` could be moved to the `symfony/contracts` too.

So `1.`, `2.` are not problem for me at all; `3.` is weird for me. `4.` not a problem anymore with `symfony/contracts`.

Another possible solution is to allow injecting custom container configurators. I can collect all closures, generate static factories for each of them and replace factory definition with it.

---

Real example: I had a need to integrate Doctrine migrations to the project, but without ORM. I had to translate all [these instructions](https://www.doctrine-project.org/projects/doctrine-migrations/en/1.7/reference/building_framework_specific_integration.html#register-console-commands) into this Symfony fluent PHP dialect of the plain PHP:
```php
    $s->set(Application::class)
        ->call('setDispatcher', [ref(EventDispatcherInterface::class)])
        ->call('setHelperSet', [ref(HelperSet::class)])
        ->call('addCommands', [[
            inline(ExecuteCommand::class),
            inline(GenerateCommand::class),
            inline(LatestCommand::class),
            inline(MigrateCommand::class),
            inline(DiffCommand::class)->args([ref(MigrationsConfigurator::class)]),
            inline(UpToDateCommand::class),
            inline(StatusCommand::class),
            inline(VersionCommand::class)
        ]]);
    $s->set(Symfony\Component\Console\Helper\HelperSet::class)->args([
        [
            'db' => inline(ConnectionHelper::class)->args([ref('doctrine_pdo')]),
            'question' => inline(QuestionHelper::class),
            'configuration' => inline(ConfigurationHelper::class)
                ->args([
                    ref('doctrine_pdo'),
                    inline(Configuration::class)->factory(
                        [ref(MigrationsConfigurator::class), 'createConfiguration']
                    )
                ])
        ]
    ]);
```
instead of copy-pasting:
```php
$s->setApplication::class)
    ->factory(function (PDO $pdo) {
        $cli = new Application();
        $cli->setHelperSet(
            new \Symfony\Component\Console\Helper\HelperSet(array(
                'db' => new \Doctrine\DBAL\Tools\Console\Helper\ConnectionHelper($pdo),
                'question' => new \Symfony\Component\Console\Helper\QuestionHelper(),
            ))
        );
        $cli->addCommands([
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\ExecuteCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\GenerateCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\LatestCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\MigrateCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\DiffCommand(
                // blah blah with $pdo
            ),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\UpToDateCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\StatusCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\VersionCommand()
        ]);
        return $cli;
    });
```

Yes, the last example is awful, but the last one at least is real PHP. 
Yes, I could create class with static method, but sometimes you're in position like "argh, I'm lazy for both variants, just, just be cool, OK?".

It is possible to add some extension point for this feature?

**Comment by stof** [comment]
The issue is that anonymous functions cannot be dumped when dumping the container, as they live in memory only, not as reusable code than can be called by the dumper.

Btw, your example of integration of doctrine/migrations is weird, as console commands should be registered in the existing CLI application, not create a separate application requiring its own bin script.

Regarding the point 3 about my remark, I don't understand it either. There is no such comment in the proposal you linked. I think it refers to a different proposal.

**Comment by unkind** [comment]
> The issue is that anonymous functions cannot be dumped when dumping the container, as they live in memory only, not as reusable code than can be called by the dumper.

Yes, but you can throw exception just like you did when people tried to configure lazy service without `ProxyManager`. It's a trade-off. 

It is possible to throw exception somewhere in `factory()` method by passing information about dumper for fail fast strategy.

> Btw, your example of integration of doctrine/migrations is weird, as console commands should be registered in the existing CLI application

It is existing. I just don't use Symfony Application directly, I have something similar to `Silly`. I can't use `DoctrineBundle`, for example, as it is coupled to the symfony-standard. This is just messy config to configure some legacy stuff.

> There is no such comment in the proposal you linked. I think it refers to a different proposal.

https://github.com/symfony/symfony/pull/23935#issuecomment-323698165

**Comment by ro0NL** [comment]
using a static factory makes more sense from DI pov. Creating incompatibility/dependencies between dumped- and built containers sounds like a bad idea IMHO.

**Comment by unkind** [comment]
> using a static factory makes more sense from DI pov. Creating incompatibility/dependencies between dumped- and built containers sounds like a bad idea IMHO.

Can you be clear, please?

**Comment by ro0NL** [comment]
im not a big fan of putting the actual factory code in DI config. It feels defeating for not actually building the service reference graph from pure definitions.

Closure factories invite to closure arguments etc. only to become more complex over time.

> Yes, but you can throw exception just like you did when people tried to configure lazy service without ProxyManager. It's a trade-off.

the lazy property is ignored when proxymanager is unavailable, so we should at least throw beforehand to not end up with a "partially" supported container that'll throw lazily, making the service graph unreliable.

**Comment by nicolas-grekas** [comment]
I've been thinking quite a lot in the past months about this and I concluded that there might be two viable alternatives to make something like this possible.

First of all, I don't consider any "SuperClosure"-like strategy viable: moving the code somewhere else means breaking DX is a very bad way, by ciphering the stack traces of any exceptions thrown from these "magically moved" closures (No, we don't want to provide source maps to fix that.)

Any alternatives should be about executing the closures in the very context where they are declared.

One option to do so could be to allow defining synthetic services using a closure factory:
`$container->set('id', $shared(true/false), function ($container) {...});`
This would be very similar to how e.g. Pimple works.
This would require extending the set() method, which currently takes an object as 2nd argument. Here, when a boolean would be passed, it would define the "shared or not" status of the return value of the now expected 3rd argument - the actual closure factory. The closure would be passed the container as argument - and by definition, it would be able to access public services only.
Calling `$container->set()` would need to be done in the kernel I suppose.

The 2nd option is more involving: at execution time, we could run the PHP-DSL file and call it with a special `ContainerConfigurator`. This configurator would then ignore any instructions there until the closure we're looking for is found. This would require some way to locate which closure is needed - eg its start/end lines - or a positional number in the flow of definitions.

**Comment by ro0NL** [comment]
that's inspiring :) IIUC it could be something like `factory(closure(function (other $o): class { }))`

closure() tracks it's own index to be stored in e.g. XML `type="closure" file="dsl.php" index="n"` correct?

**Comment by unkind** [comment]
@nicolas-grekas if I understand your second option right, it could be done like (pseudo):

```php
// Somewhere in the Symfony framework
abstract class ConfiguratorWithAnonFunctions
{
    /** @var Closure[] */
    protected $factories;

    /**
     * @param string $name
     * @param array $args
     * @return mixed
     */
    final public function runFactory(string $name, array $args = [])
    {
        if (!$this->factories[$name]) {
            throw new RuntimeException(sprintf('Factory "%s" was not found.', $name);
        }

        return $this->factories[$name](...$args);
    }

    abstract public function __invoke(ContainerConfigurator $c);
}
```

```php
// userland
return new class extends ConfiguratorWithAnonFunctions
{
    private const CLI_FACTORY = 'cli_factory';

    public function __construct()
    {
        $this->factories[self::CLI_FACTORY] =  function (
            Connection $connection,
            SchemaProvider $schemaProvider
        ) {
            $cli = new Application();
            $cli->setHelperSet(
                new \Symfony\Component\Console\Helper\HelperSet([
                    'db' => new \Doctrine\DBAL\Tools\Console\Helper\ConnectionHelper($connection),
                    'question' => new \Symfony\Component\Console\Helper\QuestionHelper(),
                ])
            );
            $cli->addCommands([
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\ExecuteCommand(),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\GenerateCommand(),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\LatestCommand(),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\MigrateCommand(),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\DiffCommand($schemaProvider),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\UpToDateCommand(),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\StatusCommand(),
                new \Doctrine\DBAL\Migrations\Tools\Console\Command\VersionCommand()
            ]);

            return $cli;
        };
    }

    public function __invoke(ContainerConfigurator $c)
    {
        $s = $c->services()
            ->defaults()
            ->autoconfigure(true)
            ->autowire(true)
            ->private();

        $s
            ->set('cli_app', \Acme\ConsoleApplication\ConsoleApplication::class)
            ->factory(ref(self::CLI_FACTORY))
            ->args(
                [ref('doctrine_connnection'), inline(SchemaProvider::class)->args([tagged('schema_provider')])]
            )
        ;
    }
};
```
And then dumper container would be just call `runFactory('cli_factory')`.
But I find it ... a bit ugly.

**Comment by unkind** [comment]
Another way is array of configuration closures:
```php
return [
    function (AnonFuncFactoryConfigurator $c) {
        $c->register('cli_factory', function (Connection $connection, SchemaProvider $schemaProvider) {
            // ...
        });

        // Another one
        $c->register(..., function () {

        })
    },
    function (ContainerConfigurator $c) {
        $s = $c->services()
            ->defaults()
            ->autoconfigure(true)
            ->autowire(true)
            ->private();

        $s
            ->set('cli_app', \Acme\ConsoleApplication\ConsoleApplication::class)
            ->factory(ref('cli_factory'))
            ->args(
                [ref('doctrine_connnection'), inline(SchemaProvider::class)->args([tagged('schema_provider')])]
            )
        ;
    }
];
```
And btw, it would be nice for custom container configurators:
```php
return [
    function (MonologContainerConfigurator $c) {
        $c->handler('file_log', function (StreamHandlerConfigurator $sc) {
            $sc->path('%kernel.logs_dir%/%kernel.environment%.log');
            $sc->level(LogLevel::DEBUG);
        };
    }
    function (DoctrineMigrationsContainerConfigurator $c) {
        // ...
        $c->setPath('%kernel.project_dir%/var/migrations/');
        $c->setTemplatePath('%kernel.project_dir%/var/migration_class_teamplate.tpl');
    },
    function (ContainerConfigurator $c) {
        $s = $c->services()
            ->defaults()
            ->autoconfigure(true)
            ->autowire(true)
            ->private();

        // rest
    }
];
```

**Comment by nicolas-grekas** [comment]
I had no specific ideas in mind about any implementations, just principles. Happy to see your ideas in return, we have some new materials to brainstorm :)

**Comment by unkind** [comment]
We can generalize the case and allow to register closures as services:
```php
return [
    function (AnonFuncConfigurator $c) {
        $c->named('foo_factory')->anonymous(function () { ... }));

        $c->tagged('monolog_processor')->anonymous(function (array $record) {
            // ...
            return $record;
        });

        $c->tagged('event_listener')->anonymous(function (Event $event) { /* ... */  });
    }

    function (MonologContainerConfigurator $c) {
        $c->processors(tagged('monolog_processor'));

        // ...
    },
;
```
The idea is the same: when compiled container faces closure-service (`Definition` could be extended with additional info like whether closure of not, config path, index [if config is array]), he loads it  (if not loaded already). Since this moment, he is just regular service: you can use it as factory, you can add it to list of services that were tagged.

**Comment by unkind** [comment]
If the last one looks acceptable, I can make a PR, but I need some kind of assurance that it is OK for all of symfony deciders, just to make sure it has high chances to be merged.

**Comment by unkind** [comment]
> moving the code somewhere else means breaking DX in a very bad way, by ciphering the stack traces of any exceptions thrown from these "magically moved" closures

I make special accent on semantical difference between closures and anonymous functions.
Closure, by definition, is an object that depends on the place within the bounds of which it was declared.
Anonymous function does not (if we don't consider using of reflection, debug_backtrace and other low-level tools).
Things become more interesting if anonymous function is pure.

I saw an RFC (withdrawn, unfortunately) with arrow functions: https://wiki.php.net/rfc/arrow_functions.
I'd like to show some examples that probably will may allow to see this issue from another the point of view and arrow syntax is just more descriptive for this purpose.

Let's just imagine that RFC was accepted, Symfony folks changed their mind and we can write:
```diff
$s->set('doctrine_pdo', Connection::class)
-    ->factory([DriverManager::class, 'getConnection'])
+    ->factory(fn(PDO $pdo) => DriverManager::getConnection($pdo))
    ->args([ref(PDO::class)]);
```
Isn't natural call of the `getConnection()` is just more straightforward? IDE is happy: semantically, it is definetely a method call and not just weird array that by coincedence is a callable. Moreover, standard IDE won't understand syntax like `[ref(FooFactory::class), 'createFoo']`, because it is just Symfony DSL and not callable from language point of view.

OK, we also have `->call()` definition artifact. 
But this artifact have special role for legacy codebase: when 2 objects are mutually dependent on each other, trick with setters helps.

Look how it can be done with anonymous functions:
```diff
$s->set('cli', Application::class)
-    ->call('addCommands', [tagged('console.command')])
+    ->postConstructorCalls(function (Application $app, iterable $commands) => $app->addCommands($commands))
    ->args([tagged('console.command')]]);
```
We again can avoid method names in strings.

After couple of months with this format, I'd say that it has a lot of artificial details for PHP config: it is not "declarative" in the same meaning as YAML/XML (so it's harder to generate/modify/analyses them) and in the same time it keeps some disadvantages of YAML/XML: IDE and plugins must be aware of rules between DI-configuration and the real classes and functions in order to provide decent level of refactoring and code-exploring tools.

So the fluent PHP-format probably should be improved with anonymous functions to be competitive with other formats (lack of "official" support of custom container configurators is a disadvantage over YAML/XML formats).

It's not that hard to find out whether anonymous function is completely context-free (no `$this`, no `use()`, no magicial `__FILE__`, `__LINE__`, no `debug_backtrace()` and other low-level tools; probably, we can just use term "like-pure-but-technically-not" function here, it is almost the same). The purpose of anonymous functions in configurations should be very straightforward: explicitly call method on the PHP-object using PHP language instead of describing the same via one more level of abstraction. 

I'm sure that "like-pure-but-technically-not" function could help to configure services way better than additional language.
Detect subset of such functions that should be allowed to be part of DI-configuration is not a rocket science as well.
Most of the job could be done with 3rd party libarties, you can just solve your part of the problem with an extension point.
And here is the question: what is so terrible in moving code of "like-pure-but-technically-not" into some private static method in compiled container?

```php
public function closurePostCtrCalls983048()
{
    try {
        $this->services['cli']->addCommands([
            new FooCommand(),
            new BarCommand($this->services['pdo'])
        ]);
    } catch (\Throwable $e) {
        throw new \LogicException(
            sprintf(
                'Closure "%s" prevents to configure "cli" service: "%s" at */app/configs/config.php:42',
                'function (Application $app, iterable $commands) => $app->addCommands($commands)',
                $e->getMessage()
            ),
            0,
            $e
        );
    }
}
```

1. What does the current level of abstraction for PHP-configs give you exactly (_explicitly call method on the PHP-object using PHP language instead of describing the same via one more level of abstraction_)?
2. What's the real horror of moving anonymous context-free functions to the different place?

**Comment by nicolas-grekas** [comment]
> What does the current level of abstraction for PHP-configs give you exactly

programmability of configuration (typically compiler passes and DI extensions).

> What's the real horror of moving anonymous context-free functions to the different place

- WTF related to special behavior of scope management
- Even bigger WTF when reading the stack traces of exceptions

**Comment by unkind** [comment]
> programmability of configuration 

Wait, but you lose it already when developer make factory class. It becomes undisclosed how object was constructed from DI-container point of view. The only difference is probably that devs will use factories more frequently. 

> typically compiler passes and DI extensions

How they live now with factories? 

> but also optimization of the service dependency graph

It remains. You can't get service inside of anonymous function without injecting it explicitly (injecting whole Container into anonymous function must be disallowed, though). So you can see on the dependency graph, that service `Foo` depends on `Bar`. 

> WTF related to special behavior of scope management

Okay, but you will throw an exception to the face with whole description of closure-factory restrictions. He'll learn it. Moreover, nowadays such behavior (moving code of the closure) is not that rare, PHP-DI, for example, exploits this feature for the same purposes. Even if he would WTF, what's wrong? Do you know how many times I WTF'ed with auto-wiring and PHP-configs (it was not explicit whether I have to continue chain or should simply write `$s->set();` with a new line. This suffering was worth it: now I can use auto-wiring. 

And restrictions on closure's scope + different stack trace (and you can help developer by descriptive exception message + provide "original" file/line) don't look impressively complex after aforementioned auto-wiring. 

I feel frustration when I have to describe all the legacy stuff with strings in PHP config with fluent PHP dialect: I'm losing too much time for nothing. I know what I'm doing, let me use these precious anonymous functions like I want. 🤨

**Comment by unkind** [comment]
@nicolas-grekas wait a minute. You allow [custom expressions](https://symfony.com/doc/current/service_container/expression_language.html):
```php
$container->autowire(Mailer::class)
    ->addArgument(new Expression('service("App\\\\Mail\\\\MailerConfiguration").getMailerMethod()'));
```
And you allow to register new expression functions:
```php

$container->register('test')
    ->setPublic(true)
    ->setClass(Logger::class)
    ->setFactory([PseudoFactory::class, 'create']) // Just returns passed argument
    ->addArgument(new Expression('blah(service("' . addslashes(Logger::class) . '"))'));
```

```php
return $this->services['test'] = \Acme\Bridges\SymfonyBridge\PseudoFactory::create(
    call_user_func(
        static function (\Acme\Logger\Logger $logger) : \Acme\Logger\Logger {
            echo 'foo', PHP_EOL;
            $logger->notice('Hello');
            return $logger;
        },
        ($this->services['Acme\Logger\Logger'] ?? $this->load('getLoggerService.php')))
);
```
(I added some new lines and indentation to show code here, but it is real one).

Can you explain why ExpressionLanguage is OK and pure PHP is not OK even if the first one allows me to achieve the same, but with additional difficulties?

**Comment by unkind** [comment]
It's even possible to allow only subset of anonymous functions that don't use more than functions `service()`, `parameter()` and arithmetic expressions like in standard version of ExpressionLanguage for DI container.

So [example from manual:](https://symfony.com/doc/current/service_container/expression_language.html) could be rewritten like
```diff
+use App\Mail\MailerConfiguration::class;
+use function Symfony\DependencyInjection\service;
+use function Symfony\DependencyInjection\parameter;

$container->autowire(Mailer::class)
-    ->addArgument(new Expression('service("App\\\\Mail\\\\MailerConfiguration").getMailerMethod()'));
+    ->addArgument(new InlinedExpression(function () {
+        return service(MailerConfiguration::class)->getMailerMethod();
+    }))
```
These functions must refer to some global variable, yes, but it will work for sure. Even if you have multiple containers, you always can guarantee that you set the right instance of container before 

Yes, I'm not fan of this code with function `Symfony\DependencyInjection\service` too, but PhpStorm has special hack for this case: https://confluence.jetbrains.com/display/PhpStorm/PhpStorm+Advanced+Metadata
So `service(Foo::class)->` can provide autocomplete for class `Foo`.

I don't suggest it, but I'm very curious how you can prefer
```php
new Expression('service("App\\\\Mail\\\\MailerConfiguration").getMailerMethod()')
```
over real PHP code
```php
return service(MailerConfiguration::class)->getMailerMethod()
```
?
In my opinion, if you allow expressions, even this subset of closures (it's not hard to check AST with PHP-Parser) looks better than `'service("App\\\\Mail\\\\MailerConfiguration").getMailerMethod()'` (my eyes are bleeding when I look at class name with 4 backslashes, I'm sure PhpStorm will never find this place if you want to rename `MailerConfiguration`, for instance.

Subset is a trade-off, but it is native code which can be supported by IDE during refactorings, by static analyzers, etc. Don't tell me that from DX perspective expression language is better. Yes, he may WTF after he finds out restrictions on closures, but with ExpressionLanguage he would WTF till end of this days. 

It shouldn't be hard to allow type hintings:
```php
    ->factory(function (MailerConfiguration $c): Mailer {
        return $c->getMailerMethod(parameter("foo"));
    })
```
to replace `service()` function.

In order to reduce number of WTFs, you can just make descriptive error message:

<pre>
Scope-dependent anonymous functions (using "use" or "$this") are not allowed for DI-factories as they cannot be properly compiled:

    ->factory(function (MailerConfiguration $c) <strong><strike>(use ($foo)</strike></strong>): Mailer {
                                                <b>↑</b>
    in /Users/acme/www/acme/src/mailer/configs/config.php:42

More information: https://symfony.com/doc/current/service_container/life_is_pain.html.
</pre>

**Comment by unkind** [comment]
@nicolas-grekas @fabpot I can make a PR for anonymous-functions-as-factories with some limitations described above. I just need to know whether you accept the idea.

It might be possible to implement it without 3rd party libraries. Just built-in tokenizer + simple parser for subset of closures:
```php
use Acme\Mail\MailerConfiguration::class;
use function Symfony\DependencyInjection\service;
use function Symfony\DependencyInjection\parameter;

// ...

->set('mailer') 
->factory(function (MailerConfiguration $c): Mailer {
    return $c->getMailerMethod(parameter('foo'));
})
->args([ref(MailerConfiguration::class)])

// or

->set('mailer') 
->factory(function (): Mailer {
    return service(MailerConfiguration::class)->getMailerMethod(parameter('foo'));
})
```
Parser can require to pass only constants (string literals and `::class` notation) to the `service()` and `parameter()` functions, so dependencies would be still statically defined.

**Comment by nicolas-grekas** [comment]
Using SuperClosure is not an option sorry. Symfony has a responsibility in carefully choosing the dependencies it forces to its users, see eg https://dunglas.fr/2018/11/about-the-dependencies-of-symfony/
In terms of software design also this is not a fit: copy/pasting code is full of side effects. Ignoring them to achieve the target feature at all costs is not something Symfony should do IMHO.

That doesn't mean we need to close this RFC - as other options have been discussed and might end up being a fit. But I'm objecting superclosure-like strategies. Let's find a correct design instead.

**Comment by ragboyjr** [comment]
@nicolas-grekas understood, and appreciate the care put into SF no doubt. It might end up being a case then that instead of supporting closures, we maybe provide an easier way to create/define service factories. My method above is my own attempt, but there could be something even better that might be supported better by the SF container.

**Comment by nicolas-grekas** [comment]
Another idea: add a `container.factory` tag (+related labeling interface for autoconfiguration).
When found, a compiler pass would register all its public methods as factories for services named after the return types they declare.

**Comment by ro0NL** [comment]
https://github.com/symfony/symfony/pull/23935 ?

**Comment by nicolas-grekas** [comment]
And so #23819 - interesting :)

**Comment by nicolas-grekas** [comment]
Another one that could be very related: #25707

**Comment by unkind** [comment]
> Another idea: add a `container.factory` tag (+related labeling interface for autoconfiguration).
When found, a compiler pass would register all its public methods as factories for services named after the return types they declare.

The point of anonymous function is to avoid separate factory class. At least, it was my original intention.

**Comment by unkind** [comment]
> copy/pasting code is full of side effects

How does expressions (`new Expression('service("App\\\\Mail\\\\MailerConfiguration").getMailerMethod()')`) differ from that?

**Comment by stof** [comment]
@unkind ExpressionLanguage in service definitions should be avoided whenever possible. It is indeed very bad for DX (and does not play well with most newer features of Symfony DI, as it relies on accessing public services for `service()` IIRC)

**Comment by ragboyjr** [comment]
Going off of one of the original ideas from @unkind, not sure if this even that good of an idea, but it could spark some more conversation....

What if we just added a new php file `config/service-factories.php` (or whatever), and it's just a map of service ids with a closure factories.

```php
return [
  AcmeService::class => function() {
    // works exactly like a normal factory function
  }
]
```

Then the di container could easily access these functions at compile time and work out the args properly for auto-wiring and whatnot, and at runtime, those closures can be loaded at runtime. If any error occurs, then all of the stack trace information would be ok.

**Comment by nicolas-grekas** [comment]
@ragboyjr I think that's exactly what `ServiceProviderInterface` would allow, in a cleaner way to me.

**Comment by ragboyjr** [comment]
@nicolas-grekas Forgive me, but can you point me to the discussion or documentation around ServiceProviderInterface?

**Comment by nicolas-grekas** [comment]
I only have the PR and its linked issue: #25707

**Comment by ragboyjr** [comment]
@nicolas-grekas neato! one thing that isn't clear from the service provider interface if the service factories are able to accept parameters from the container, and if they are auto wired or not?

**Comment by nicolas-grekas** [comment]
No parameters, but PSR 11 allows scalars/arrays.
Autowiring is undefined yet.

**Comment by unkind** [comment]
> What if we just added a new php file config/service-factories.php (or whatever), and it's just a map of service ids with a closure factories.

It kills the idea for me: the idea is to have very cohesive DI-config so I can define service (and such factory is most cases is part of DI-definition for me) in the single file. If I need to go to another one, I can just make regular factory for this service.

---

`ServiceProviderInterface` just allows to define a service locator, right? I'm a bit confused: how does it help?

**Comment by unkind** [comment]
You probably know [wasm](https://github.com/wasmerio/php-ext-wasm/blob/master/examples/simple.rs) project where they can convert anonymous function to WebAssembly code and "restore" it in different language like Rust. Yes, the anonymous function must be simple, but I treat anonymous function factory in the same way: it is a tiny instruction that says how service object must be instantiated. It's OK to modify some aspects of the code, the only important thing is to transfer the instruction, algorithm by itself. 

> In terms of software design also this is not a fit: copy/pasting code is full of side effects.

"Copy/pasting" is just a buzzword with a negative connotation.

So you say "copy/pasting" is full of side effects.
Static anonymous function is an isolated piece of code.
Side effect by definition is when function changes something outside of its scope.
But how static anonymous function can be full of side effects? 
I'm afraid, such function probably cannot be called "static" anymore.

**Comment by nicolas-grekas** [comment]
I'd suggest sending poc code at this point. The discussion won't move forward otherwise. That's not a promise it will convince reviewers (but I can promise no-code won't convince anyone for sure).

I've raised three objections before:
1. programmability of configuration
2. WTF related to special behavior of scope management
3. Even bigger WTF when reading the stack traces of exceptions/errors

1. is challenged, I can get why, show me the code to prove the claims
2. can be dealt with by forbidding these at compile time, show me the code also
3. I doubt it's trivial, but I'd be happy to be wrong, show me the code too.

(for the poc, SuperClosure can be used I'd say if it makes it easier, then we might decide to not rely on it).
To have the poc, why

**Comment by unkind** [comment]
> I'd suggest sending poc code at this point. The discussion won't move forward otherwise.

Try to understand me too, I tried it for a few times in the past: #11968, #12115 (and many times tried to raise it in similar threads), but it seems like the concept of anonymous function factory by itself is the issue for @symfony deciders. That's why I want to find some positive feedback regarding the idea itself and only then I can make a pull request.

> programmability of configuration (typically compiler passes and DI extensions but also optimization of the service dependency graph)

First of all, looking at our discussion, I think it was my offtopic in the first place, because the very same abilities you have with a regular factory.
Yes, I want to make it easier to writing configuration using IDE, natural language abilities, etc.
But it doesn't make anonymous factories bad per se. 

Second, your argument is tricky: who says that such config-based way to run a method gives any benefit (that my example from above):

```diff
$s->set('doctrine_pdo', Connection::class)
-    ->factory([DriverManager::class, 'getConnection'])
+    ->factory(fn(PDO $pdo) => DriverManager::getConnection($pdo))
    ->args([ref(PDO::class)]);
```
? We know the fact: we use `DriverManager::getConnection()` method if you use `[DriverManager::class, 'getConnection']` notation. But what exactly does it give? Also I may say that this config-notation will mess up my method renaming refactoring from time to time.

Can you provide some examples when old approach is better than new one?

> WTF related to special behavior of scope management

PHP doesn't make strict difference between closures and anonymous functions, but I don't see anything special in last ones.

I can show example from my project, it looks dumb, but it works ([full code](https://gist.github.com/unkind/8cafd0da4f13c709c3db7d1bd3ba255f)) (AFAIK, SuperClosure uses the same library as main adapter for parser):
```php
$ast = new Ast($closureNode->stmts);
$ast->traverse(
    [
        function (Variable $node) use ($reflection) {
            if ($node->name === 'this') {
                throw new LogicException(
                    sprintf(
                        'Closure must not use $this variable in %s:%d.',
                        $reflection->getFileName(),
                        $reflection->getStartLine()
                    )
                );
            }
        },

        function (ClosureUse $node) use ($reflection) {
            throw new LogicException(
                sprintf(
                    'Closure must not use parent scope variables in %s:%d.',
                    $reflection->getFileName(),
                    $reflection->getStartLine()
                )
            );
        }
    ]
);
```

It's possible to allow only subset of language features to make it as "safe" as expression language is.

> I doubt it's trivial to fix, but I'd be happy to be wrong, show me the code too.

What's the problem with it? You mean that line:
```
#1 /Users/foo/www/bar/test.php(33): {closure}()
```
?
I think most of devs just ignore most of stack trace that goes across DependencyInjection component. :)

---

I don't have my own stats regarding WTFs by using dumped/generated code, but it would be cool if @mnapoli can say something about this, he implemented the same feature in PHP-DI long time ago IIRC.

---

I understand that Symfony doesn't want to pull some 3rd party dependencies like SuperClosure or PhpParser, but Symfony can at least make public contract `ClosureDumper` and use it in the `DependencyInjection`, so everyone (?) will be happy.

**Comment by mnapoli** [comment]
> I don't have my own stats regarding WTFs by using dumped/generated code, but it would be cool if @mnapoli can say something about this, he implemented the same feature in PHP-DI long time ago IIRC.

Right, closures are compiled in PHP-DI since v6, released 1.5y ago. As a maintainer, I haven't seen any issue with this (AFAIR). As a user, I am very happy with the feature, haven't seen any problem either. To me the conclusion I can draw from my experience is that it works reliably.

Here are the tests concerning that specific bit:

https://github.com/PHP-DI/PHP-DI/blob/00a5eb0f0dd94deabb71acfbf44c37b80d53d0a1/tests/IntegrationTest/Definitions/FactoryDefinitionTest.php#L431-L553

And here is the piece of code that turns the Closure into a string (code):

https://github.com/PHP-DI/PHP-DI/blob/00a5eb0f0dd94deabb71acfbf44c37b80d53d0a1/src/Compiler/Compiler.php#L346-L378

**Comment by unkind** [comment]
By the way, I actively use closures for many configurations that containing `$this`. In this case I just require to have DI-service for this class, instantiate it and bind my copied closure. This is very handy.
